macrosynergy/visuals/README.md

```diff
@@ -0,0 +1,186 @@
+# macrosynergy.visuals
+
+## Description
+
+The `macrosynergy.visuals` subpackage contains functions for visualizing quantamental data, and to provide `view.*` functions for the `macrosynergy` package.
+
+The functionality is built around the ability to create quick generic plots of data, and to provide a framework for creating custom plots.
+Simplicity, speed, consistency and flexibility are the main goals.
+The core visualization functions are built around `matplotlib`, with goals of mimicking the rich and dataframe-friendly nature of `seaborn`.
+
+## Functionality
+
+The `macrosynergy.visuals` subpackage features the following functionality:
+
+### User-facing functions:
+
+- `view.*` functions: A set of preset methods to view data in commonly used contexts.
+
+  - `timelines()`: A function for viewing time-series data. Uses `FacetPlot.lineplot()` and `LinePlot` internally.
+  - `availability()`: A function for viewing data availability. Uses `Heatmap` internally.
+  - `correlation()`: A function for viewing correlation matrices. Uses `Heatmap` internally.
+  - `metrics()`: A function for viewing metrics. Uses `Heatmap` internally.
+  - `ranges()`: A function for viewing ranges. Uses `BoxPlot` and `BarPlot` internally.
+  - `distribution()`: A function for viewing distributions. Uses `FacetPlot.boxplot()` and `BoxPlot` internally.
+  - `reg_scatter()`: A function for viewing scatter plots. Uses `FacetPlot.scatterplot()` and `ScatterPlot` internally.
+
+### Backend functions:
+
+- `FacetPlot`: A class for creating facet plots of data.
+
+  - `.lineplot()`: A method for creating line plots.
+  - `.scatterplot()`: A method for creating scatter plots.
+  - `.from_subplots()`: A method for copying and arranging a list of individual subplots (of any type) into a facet plot.
+
+- `LinePlot`: A class for a single line chart.
+- `ScatterPlot`: A class for a single scatter chart.
+- `BoxPlot`: A class for a single box plot.
+- `BarPlot`: A class for a single bar chart.
+- `Heatmap`: A class for a single heatmap.
+
+## Design principles
+
+The interfaces follow a object-oriented approach, as opposed to a functional approach.
+
+The base class for all of the "plotter" classes is `Plotter` - `macrosynergy.visuals.Plotter`.
+This class is not intended to be used directly, but rather to be subclassed by other classes to inherit common functionality.
+The base class implements dataframe reduction and filtering, functions. In the future, this can be extended to allow handling of larger-than-memory dataframes.
+
+Another important feature is the metaclass the `Plotter` is based on - `macrosynergy.visuals.PlotterMetaClass`.
+The metaclass implements two important features:
+
+- `argvalidation`: A decorator for validating arguments passed to methods.
+- `argcopy`: A decorator for copying arguments passed to methods, preventing the original arguments from being modified.
```

Comment:
Clever on `argcopy` but do we need to allow for an override, in case of large DataFrame's where it might not be memory desirable to copy?

Reply:
True, I am considering limiting this to the class instantiation...